internal/pkg/earpointer/analysis_test.go

```diff
@@ -651,3 +644,303 @@ func TestUnimplemented(t *testing.T) {
+	}
+}
+
+func TestGoDefer(t *testing.T) {
+	code := `package p
+	func g(i *int) {
+		go func() {
+ 			print(*i)
+		}()
+		defer func(k *int) {
+ 			print(*k)
+		}(i)
+	}
+	`
+	/*
+		func g(i *int):
+		0:                                       entry P:0 S:0
+			t0 = new *int (i)                    **int
+			*t0 = i
+			t1 = make closure g$1 [t0]           func()
+			go t1()
+			t2 = *t0                             *int
+			defer g$2(t2)
+			rundefers
+			return
+		1:                                       recover P:0 S:0
+			return
+	*/
+	state, err := runCode(code)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Free variable g$1.i is unified with g.t0 due to closure binding;
+	// Arguments g$2.k and g.i are unified due to the defer call.
+	want := "{g$1.i,g.t0}: --> g.t2, {g$1.t0}: [], {g$2.k,g.i,g.t2}: [], " +
+		"{g.t1}: []"
+	if got := state.String(); got != want {
+		t.Errorf("got: %s\n want: %s", got, want)
+	}
+}
+
+func TestCallBuiltin(t *testing.T) {
+	code := `package p
+	func f(s []*int, x *int, d1 []*int) {
+		d2 := append(s, x)
+		copy(d1, s)
+		print(d1, d2)
+	}
+	`
+	/*
+		func f(s []*int, x *int, d1 []*int):
+		0:                                             entry P:0 S:0
+			t0 = new [1]*int (varargs)                 *[1]*int
+			t1 = &t0[0:int]                            **int
+			*t1 = x
+			t2 = slice t0[:]                           []*int
+			t3 = append(s, t2...)                      []*int
+			t4 = copy(d1, s)                           int
+			t5 = print(d1, t3)                         ()
+			return
+	*/
+	state, err := runCode(code)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// f.s and f.d1 are separate due to the "copy".
+	// f.d1 contains a field pointing at f.x due to the "copy".
+	// f.s, f.t2 and f.r3 ("d2") are unified due to the "append".
+	want := "{f.d1[.],f.t1}: --> f.x, {f.d1}: [0->f.t1], " +
+		"{f.s,f.t0,f.t2,f.t3}: [0->f.t1], {f.x}: []"
+	if got := state.String(); got != want {
+		t.Errorf("got: %s\n want: %s", got, want)
+	}
+}
+
+func TestVariadicCall(t *testing.T) {
+	code := `package p
+	func g(ks ...*int) *int {
+ 		return ks[0]
+	}
+	func f(a *int, b *int) *int {
+ 		return g(a, b)
+	}
+	`
+	/*
+		func f(a *int, b *int) *int:
+		0:                                                     entry P:0 S:0
+			t0 = new [2]*int (varargs)                         *[2]*int
+			t1 = &t0[0:int]                                    **int
+			*t1 = a
+			t2 = &t0[1:int]                                    **int
+			*t2 = b
+			t3 = slice t0[:]                                   []*int
+			t4 = g(t3...)                                      *int
+			return t4
+
+		func g(ks ...*int) *int:
+		0:                                                     entry P:0 S:0
+			t0 = &ks[0:int]                                    **int
+			t1 = *t0                                           *int
+			return t1
+	*/
+	state, err := runCode(code)
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := "{f.a,f.t4,g.t1}: [], {f.b}: [], {f.t0,f.t3,g.ks}: " +
```

Comment:
I am pleasantly surprised we have enough precision to split f.a and f.b. What enabled this?

Reply:
Again SSA does some useful work here by first creating an array and storing arguments "a" and "b" into the array. Now f.a and f.b are the elements of this array (which is unified with the function argument).

  t0 = new [2]*int (varargs)                         *[2]*int
  t1 = &t0[0:int]                                    **int
  *t1 = a
  t2 = &t0[1:int]                                    **int
  *t2 = b

Reply:
How are
```
  t1 = &t0[0:int] **int
  t2 = &t0[1:int] **int
```
separate? Are we keeping different 0 and 1 indices? If so where?